shared/reducers/notifications.js

```diff
@@ -1,74 +1,52 @@
-import {isMobile} from '../constants/platform'
+import {chatTab, folderTab} from '../constants/tabs'
 
-import type {NotificationKeys, NotificationAction, BadgeType, MenuNotificationState} from '../constants/notifications'
+const initialState: Constants.State = new Constants.StateRecord()
 
-type State = {
-  keyState: {
-    [key: NotificationKeys]: boolean,
-  },
-  menuBadge: BadgeType,
-  menuBadgeCount: number,
-  menuNotifications: MenuNotificationState,
-}
+const _updateWidgetBadge = (s: Constants.State): Constants.State => {
+  let widgetBadge = 'regular'
+  // $FlowIssue getIn
+  if (s.getIn(['keyState', 'kbfsUploading'])) {
+    widgetBadge = 'uploading'
+  } else if (s.desktopAppBadgeCount) {
+    widgetBadge = 'badged'
+  }
 
-const initialState = {
-  keyState: {},
-  menuBadge: 'regular',
-  menuBadgeCount: 0,
-  menuNotifications: {
-    chatBadge: 0,
-    deviceBadge: 0,
-    folderBadge: 0,
-    peopleBadge: 0,
-  },
+  return s.set('widgetBadge', widgetBadge)
 }
 
-export default function (state: State = initialState, action: NotificationAction): State {
+export default function (state: Constants.State = initialState, action: Constants.Actions): Constants.State {
   switch (action.type) {
     case CommonConstants.resetStore:
-      return {...initialState}
-    case Constants.badgeApp:
-      // $ForceType
-      const badgeAction: BadgeAppAction = action
-
-      if (badgeAction.error) {
-        return state
-      }
-      let keyState = {
-        ...state.keyState,
-      }
-
-      keyState[badgeAction.payload.key] = badgeAction.payload.on
-      const menuNotifications = {...state.menuNotifications}
-
-      let menuBadge = 'regular'
-      if (keyState.kbfsUploading) {
-        menuBadge = 'uploading'
-      } else if (keyState.newTLFs || keyState.chatInbox) {
-        menuBadge = 'badged'
-      }
-
-      if (badgeAction.payload.key === 'newTLFs') {
-        menuNotifications.folderBadge = badgeAction.payload.count || 0
-      } else if (badgeAction.payload.key === 'chatInbox') {
-        menuNotifications.chatBadge = badgeAction.payload.count || 0
-      }
-
-      // Menu badge is chat only currently
-      const menuBadgeCount = isMobile
-        ? menuNotifications.chatBadge
-        : Object.keys(menuNotifications).reduce((total, n) => total + menuNotifications[n], 0)
-
-      return {
-        ...state,
-        keyState,
-        menuBadge,
-        menuBadgeCount,
-        menuNotifications,
-      }
+      return initialState
+    case 'notifications:receivedBadgeState': {
+      const {conversations, newTlfs, rekeysNeeded} = action.payload.badgeState
+
+      const navBadges = state.get('navBadges').withMutations(n => {
+        const totalMessages = (conversations || []).reduce((total, c) => total + c.UnreadMessages, 0)
+        n.set(chatTab, totalMessages)
+        n.set(folderTab, newTlfs + rekeysNeeded)
+      })
+
+      // $FlowIssue withMutations
+      let newState = state.withMutations(s => {
```

Comment:
do we really need to use `withMutations`?
seems like there's just more context needed to use withMutations (e.g. you can only use certain methods)

this would be about the same as:

```js
let newState = state
  .set('navBadges', navBadges)
  .set('desktopAppBadgeCount', navBadges.reduce((total, val) => total + val, 0))
  .set('mobileAppBadgeCount', navBadges.get(chatTab, 0))
```